client/src/main/java/io/pravega/client/netty/impl/SessionHandler.java

```diff
@@ -0,0 +1,241 @@
+import io.pravega.shared.protocol.netty.ConnectionFailedException;
+import io.pravega.shared.protocol.netty.Reply;
+import io.pravega.shared.protocol.netty.ReplyProcessor;
+import io.pravega.shared.protocol.netty.WireCommands;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class SessionHandler extends ChannelInboundHandlerAdapter implements AutoCloseable {
+
+    private static final int SESSION_DISABLED = -1;
+    private final String connectionName;
+    private final AtomicReference<Channel> channel = new AtomicReference<>();
+    private final AtomicReference<ScheduledFuture<?>> keepAliveFuture = new AtomicReference<>();
+    private final AtomicBoolean recentMessage = new AtomicBoolean(false);
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final AppendBatchSizeTracker batchSizeTracker;
+    private final ReusableFutureLatch<Void> registeredFutureLatch = new ReusableFutureLatch<>();
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final ConcurrentHashMap<Integer, ReplyProcessor> sessionIdReplyProcessorMap = new ConcurrentHashMap<>();
+    private final AtomicBoolean disableSession = new AtomicBoolean(false);
+
+    public SessionHandler(String connectionName, AppendBatchSizeTracker batchSizeTracker) {
+        this.connectionName = connectionName;
+        this.batchSizeTracker = batchSizeTracker;
+    }
+
+    /**
+     * Create a session on existing connection.
+     * @param session Session.
+     * @param rp ReplyProcessor for the specified session.
+     * @return Client Connection object.
+     */
+    public ClientConnection createSession(final Session session, final ReplyProcessor rp) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Preconditions.checkState(!disableSession.get(), "Ensure sessions are enabled.");
+        log.info("Creating Session {} for Endpoint {}. The current Channel is {}.", session.getSessionId(), connectionName, channel.get());
+        if (sessionIdReplyProcessorMap.put(session.getSessionId(), rp) != null) {
+            throw new IllegalArgumentException("Multiple sessions cannot be created with the same Session id " + session.getSessionId());
+        }
+        return new ClientConnectionImpl(connectionName, session.getSessionId(), batchSizeTracker, this);
+    }
+
+    /**
+     * Create a {@link ClientConnection} where sessions are disabled. This implies that there is only one session on the underlying
+     * network connection.
+     * @param rp  The ReplyProcessor.
+     * @return Client Connection object.
+     */
+    public ClientConnection createConnectionWithSessionDisabled(final ReplyProcessor rp) {
```

Comment:
Where do we need this?

Reply:
this is invoked when the user user of connection factory invokes `establishConnection` without passing the `Flow` id. Currently the controller (SegmentHelper) is using this api.